README.md
# CDN Connect API Ruby Client, v0.3.1

CDN Connect makes it easier to manage production assets for teams of developers and designers, all while serving files from a fast content delivery network. Features include image optimization, resizing, cropping, filters, changing output formats, convert to WebP image format, etc. The CDN Connect API Ruby Client makes it easier to upload files and interact with the API with only a few lines of code.

[View the full CDN Connect API documentation](http://api.cdnconnect.com/)

## Install

    $ gem install cdnconnect-api

[RubyGems.org: cdnconnect-api](https://rubygems.org/gems/cdnconnect-api)


## Setup the API Client

First step is to create an api client instance which will be used to connect to your CDN Connect app. The required options are `app_host` and `api_key`.

#### App Host
The CDN Connect App host includes your app subdomain and the `cdnconnect.com` domain. For example, `demo.cdnconnect.com` is a CDN Connect app host. The app host should not include `https://`, `http://` or a URL path such as `/images`.

#### API Key

Most interactions with CDN Connect APIs require users to authorize applications via OAuth 2.0. An API Key can be created for a specific app within your CDN Connect's account. Sign into your account and go to the "API Key" tab for the app you want to interact with. Next click "Add API Key" and use this value when creating a new API client within the code. The API Key can be revoked by you at any time and numerous keys can be created.


#### Example API Client

    require 'cdnconnect_api'

    api_client = CDNConnect::APIClient.new(:app_host => 'YOUR_APP.cdnconnect.com', 
                                           :api_key => 'YOUR_API_KEY')


## Upload Files

Upload a file or multiple files from a local machine to a folder within a CDN Connect app. The `upload` method provides numerous ways to upload files or files, to include recursively drilling down through local folders and uploading only files that match your chosen extensions. If any of the folders within the upload path do not already exist then they will be created automatically.

Below are the possible parameters for the `upload` method. You must set `destination_path` and use one of the options to select where the source files are uploaded from.

 - `destination_path` : The URL of the CDN Connect folder to upload to. If the destination folder does not already exist it will automatically be created.
 - `source_file_path` : A string of a source file's local path to upload to the destination folder. If you have more than one file to upload it'd be better to use `source_file_paths` or `source_folder_path` instead.
 - `source_file_paths` : A list of a source file's local paths to upload. This option uploads all of the files to the destination folder. If you want to upload files in a local folder then `source_folder_path` option may would be easier than listing out files manually.
 - `source_folder_path` : A string of a source folder's local path to upload. This will upload all of the files in this source folder to the destination url. By using the `valid_extensions` parameter you can also restrict which files should be uploaded according to extension.
 - `valid_extensions` : An array of valid extensions which should be uploaded. This is only applied when the `source_folder_path` options is used. If nothing is provided, which is the default, all files within the folder are uploaded. The extensions should be in all lower case, and they should not contain a period or asterisks. Example `valid_extensions => ['js', 'css', 'jpg', jpeg', 'png', 'gif', 'webp']`
 - `recursive_local_folders` : A true or false value indicating if this call should recursively upload all of the local folder's sub-folders, and their sub-folders, etc. This option is only used when the `source_folder_path` option is used.
 - `destination_file_name` : The name which the uploaded file should be renamed to. By default the file name will be the same as the file being uploaded. The `destination_file_name` option is only used for a single file upload, it does not work for multiple file requests.
 - `queue_processing` : A true or false value indicating if the processing of the data should be queued or processed immediately. A response with "queued_processing" will be faster because the resposne doesn't wait on the system to complete processing the data. However, because an queued processing response does not wait for the data to complete processing then the response will not contain any information about the data which was just uploaded. Use queued processing only if you do not need to know the details of the upload. Additionally you can use the `webhook_url` to post back the uploads details once it's processed. Default is true.
 - `webhook_url` : A URL which the system should `POST` the response to. This works for both immediate processing or queued processing calls. The data sent to the `webhook_url` will be the same as the data that is responded in a synchronous response, and is sent within the `data` parameter. The format sent can be in either `json` or `xml` by using the `webhook_format` parameter. By default there is no webhook URL.
 - `webhook_format` : When a `webhook_url` is provided, you can have the data formatted as either `json` or `xml`. The defautl format is `json`.

### Upload One File: `source_file_path`

Use this option if you simply want to upload just one file. If you have many files to upload we recommend using either `source_file_paths` or `source_folder_path`.

    response = api_client.upload(:destination_path => '/images', 
                                 :source_file_path => '/Users/Ellie/Pictures/meowzers.jpg')
    
### Upload A List Of Files: `source_file_paths`
    
Specify a list of local files that should be uploaded to an app folder. Use this option if you want to manually select which files should be uploaded. Use the `source_folder_path` option if you want to easily upload all of the files
in a folder.
    
    response = api_client.upload(:destination_path => '/images/kitty', 
                                 :source_file_paths => [
                                    '/Users/Ellie/Pictures/furball.jpg',
                                    '/Users/Ellie/Pictures/smuckers.jpg',
                                    '/Users/Ellie/Pictures/socks.jpg'
                                 ])
    
### Upload All Of The Files In The Folder: `source_folder_path`
    
All files within the local `Pictures` folder will be uploaded. Additionally, by default all files within its subfolders will also be uploaded. Refer to the `recursive_local_folders` parameter if you do not want to recursively upload files in subfolders. 
    
    response = api_client.upload(:destination_path => '/images/', 
                                 :source_folder_path => '/Users/Ellie/Pictures/')


## Get File or Folder Information

Both files and folders are considered "objects", and object data contains information stating if it is a file or a folder. A folder can contain many sub-folders, and many files, and a file is contained by a folder. The concept of files and folders is no different than how your computer handles them, and their hierarchy is what builds the URL. Getting information about a file or a folder both use `get_object`.


#### Get File Information

    response = api_client.get_object(:path => '/images/spacewalk.jpg')


#### Get Folder's Basic Information

    response = api_client.get_object(:path => '/images')


#### Get Folder's Sub-Files and Sub-Folders

    response = api_client.get_object(:path => '/images', :files => true, :folders => true)


## Rename File or Folder

Renames a file or folder, which are both also known as an object.

    response = api_client.rename_object(:path => '/images/tv-shows/night-rider.jpeg',
                                        :new_name => 'knight-rider.jpg')


## Create A Folder Path

Creates a folder structure according to the path provided. If any of the folders do not already exist they will be created. The response contains data for every folder in the path, new and existing. The feature of creating the path automatically is also available when uploading files.

In the example below, if the folders `images` or `movies` did not already exist with the CDN Conenct app then they would automatically be created.

    response = api_client.create_path(:path => '/images/movies')


## API Response

HTTP responses will be formatted in json, but the library takes the HTTP response and decodes into a hash for the `APIResponse` class. The `APIResponse` class is used to simpilfy things by using helper functions to read response data. Responses from the API are all structured the same way, and this class is used as a small wrapper to make it easier to get data from it. Note: Be sure to set the :queue_processing => false upload option if you would like to get the upload info the upload's response.


 - `files` : `array` : A list of all the files that were uploaded. Each file in the array is a hash.
 - `object`: `hash` : Can be either a file or folder, or the first file in the `files` array.
 - `msgs ` : `array` : An array of messages, and each message is a hash. Example message within the `msgs` array: `{"text" => "info about the message", "status" => "error"}`
 - `is_success` : `bool` : Successful API call, the response should contain the data your looking for.
 - `is_error` : `bool` : Unsuccessful API call. Could be a client error (400) or a server error (500).
 - `is_client_error` : `bool` : Unsuccessful API call due to a client error. Review the `msgs` array for more info.
 - `is_bad_request` : `bool` : Unsuccessful API call due to sending invalid data. Review the `msgs` array for more info.
 - `is_unauthorized` : `bool` : Unsuccessful API call due to not being authorized.
 - `is_not_found` : `bool` : Unsuccessful API call because the resource does not exist.
 - `is_server_error` : `bool` : Unsuccessful API call because server is having issues (its also possible, but hopefully you'll never see this).


#### Example Upload Response

    if response.is_success

        # Be sure to set the :queue_processing => false upload option if you would like to get the upload info like the response below
        for file in response.files
            puts "Uploaded " + file["name"]
        end

    end


#### Example Get Object Response

    if response.is_success

        puts "Got object " + response.object["name"]

    end


#### Example Error Response

    if response.is_error

        puts "CDN Connect Error"

        for msg in response.msgs
            puts msg["status"] + ": " + msg["text"]
        end

    end

Note that this HTTP response will be parsed and can be easily read using the APIResponse. Be sure to view the [API documentation](http://api.cdnconnect.com/) describing what each response object will contain depending on the API resource.


## Support

Please [report any bugs](https://github.com/cdnconnect/cdnconnect-api-ruby/issues) on this project's issues page. Additionally, please don't hesitate to contact us and/or submit a pull request with any ideas you may have to improve the service. We will continue to improve upon and build out the API in order to bring more value to you and your projects.



CHANGELOG.md
# 0.3.1

 * Added timeout of 1.5 seconds after an upload error


# 0.3.0

* Reuse the same upload url if there is an upload error
* Add an upload_id parameter to the upload url so it 
  always gets a new one when requested (cannot be querystring!)
* Added Ruby Logger: 
  http://www.ruby-doc.org/stdlib-1.8.7/libdoc/logger/rdoc/Logger.html
* queue_processing is the default
* build_post_data() method now uses options={} arg


# 0.2.5

* Relaxed signet dependency to version 0.4.5 so it works
  with Ruby 1.8.7


# 0.2.4

* Renamed "async" parameter to "queue_processing"
* Added "destination_file_name" parameter so you can specify what
  a file's name will be once it's uploaded


# 0.2.3

* Authentication header no longer required on upload POST.
  The upload url contains encrypted information, can only
  be used once, and if its not used it expires.


# 0.2.2

* Correctly send webhook_url and webhook_format


# 0.2.1

* Support for receiving a files and folders within a folder


# 0.2.0

* Group multiple files into one request
* Upload an entire folder and recursively drill down and upload
  all of its sub-folder's files
* Upload multiple files with an array of local file paths
* Set an async option so processing the upload is asynchronous
  allowing the upload response to be faster
* Set a webhook_url option
* Add 'get_object' method
* Add 'rename_object' method
* Add 'delete_object' method
* Add 'create_path' method
* Add 'files' and 'object' properties to the APIResponse object
* Removed ability to upload via app_id and obj_id


# 0.1.2

* Remove XML from data format possibilities. API supports XML
  but there is no need for this library to support both.
  

# 0.1.1

* Client needs to specify the version number in the path
* Library does not assume /v1/ in the path
* Added debug option to print out debugging info


# 0.1.0

* Have the client call the exact API path they need data from.
* Added http verb methods for GET/POST/PUT/DELETE API requests.
* get() method is no longer used to receive object information only.
* Initial alpha release


# 0.0.2

* Documentation updates


# 0.0.1

* Initial dev release

test/http_response.rb

class CDNTestHttpResponse

    def initialize(status, body)
      @status = status
      @body = body
      return self
    end

    def status
        @status
    end

    def body
        @body
    end
    
end

test/testfiles/subfolder1/subfolder1a/script.js
alert("Hello, I'm Dr. Tobias Funke");

sample.rb
# If you haven't already, install the cdnconnect-api gem
# gem install cdnconnect-api

require 'cdnconnect_api'



# Your CDN Connect App Config
APP_HOST = 'demo.cdnconnect.com'
API_KEY = '1bPz6lscwNQjs5cBJEw3Pi4olaQccorpnkbSSSFNf'



# Create an API Client
api_client = CDNConnect::APIClient.new(:api_key => API_KEY, 
                                       :app_host => APP_HOST,
                                       :log_device => STDOUT, # log_device could also be a file path
                                       :log_sev_threshold => Logger::DEBUG)



puts "Upload one file (default uses queue_processing)"
response = api_client.upload(:destination_path => '/sample/one-file', 
                             :source_file_path => './test/testfiles/DSCN0251.JPG')

if response.is_success
  puts "File uploaded and processing will be queued"
end



puts "Upload multiple files to the same folder using an array of file paths"
response = api_client.upload(:destination_path => '/sample/multiple-files', 
                             :source_file_paths => [
                                './test/testfiles/IMAG0355.jpg',
                                './test/testfiles/IMAG4501.jpg'
                              ],
                              :queue_processing => false)

if response.is_success
  
  for file in response.files
    puts " - #{file["name"]}"
    for msg in file["msgs"]
      puts "    - #{msg["text"]}"
    end
  end

end


puts "Upload all files within a folder recursively, keep the same path structure"
response = api_client.upload(:destination_path => '/sample/entire-folder', 
                             :source_folder_path => './test/testfiles',
                             :queue_processing => false)

if response.is_success
  
  for file in response.files
    puts " - #{file["name"]}"
    for msg in file["msgs"]
      puts "    - #{msg["text"]}"
    end
  end

end



puts "Upload certain files with specific file extensions within a folder recursively"
response = api_client.upload(:destination_path => '/sample/folder-specific-extensions', 
                             :source_folder_path => './test/testfiles',
                             :valid_extensions => ['js', 'css', 'jpg'],
                             :queue_processing => false)

if response.is_success

  for file in response.files
    puts " - #{file["name"]}"
    for msg in file["msgs"]
      puts "    - #{msg["text"]}"
    end
  end

end



puts "Upload multiple files within a folder path not recursively"
response = api_client.upload(:destination_path => '/sample/folder-no-recursion', 
                             :source_folder_path => './test/testfiles',
                             :recursive_local_folders => false,
                             :queue_processing => false)

if response.is_success

  for file in response.files
    puts " - #{file["name"]}"
    for msg in file["msgs"]
      puts "    - #{msg["text"]}"
    end
  end

end



puts "Upload multiple files, queue the processing of the data, then send response to webhook"
# Response will not contain uploaded file data but uploads will be quicker
# In this example, when the files are done processing the data will be sent via POST
# to the webhook_url
response = api_client.upload(:destination_path => '/sample/queue_processing', 
                             :source_folder_path => './test/testfiles',
                             :queue_processing => true,
                             :webhook_url => 'http://yousite.com/webhook')

if response.is_success
  puts "Queued processing upload completed"
end



puts "Get a folder's basic information only"
response = api_client.get_object(:path => '/sample')

if response.is_success
  puts "Got folder #{response.object["name"]}"
end



puts "Get a folder's sub-files and sub-folders"
response = api_client.get_object(:path => '/sample/entire-folder', :files => true, :folders => true)

if response.is_success
  puts "Got folder #{response.object["name"]} and its #{response.object["files"].length} files and #{response.object["folders"].length} folders"
end



puts "Get a file's information"
response = api_client.get_object(:path => '/sample/one-file/DSCN0251.JPG')

if response.is_success
  puts "Got folder #{response.object["name"]}"
end



puts "Create a path"
response = api_client.create_path(:path => '/sample/creating/a/new/path/in/one/request')

if response.is_success
  puts "Renamed #{response.results["old_name"]} to #{response.results["new_name"]}"
end



puts "Rename a file"
response = api_client.rename_object(:path => '/sample/one-file/DSCN0251.JPG', 
                                    :new_name => 'DSCN0251-renamed.jpg')

if response.is_success
  puts "Renamed #{response.results["old_name"]} to #{response.results["new_name"]}"
end


puts "Upload a file so it can be deleted"
response = api_client.upload(:destination_path => '/sample/delete-file', 
                             :source_file_path => './test/testfiles/DSCN0251.JPG',
                             :queue_processing => false
                             )
if response.is_success
  puts "Test file uploaded so it can be deleted in the next test"
end

puts "Delete the uploaded test file"
response = api_client.delete_object(:path => '/sample/delete-file/DSCN0251.JPG')
if response.is_success
  puts "File deleted"
end




lib/cdnconnect_api/response.rb
# Copyright 2013 CDN Connect
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'json'

module CDNConnect

  ##
  # Used to simpilfy things with helper functions to read response 
  # data. Responses from the API are all structured the same way, and
  # this class is used as a wrapper to make it easier to read 
  # resposne data.
  class APIResponse

    def initialize(http_response=nil)
      @http_response = http_response
      if http_response != nil
        @status = http_response.status
        @data = nil
        @body = nil
      else
        @status = 0
        @data = {"results" => {}, "msgs" => []}
        @body = nil
      end
      return self
    end

    
    ##
    # A list of all the files that were uploaded. Each file in the array is a hash.
    #
    # @return [array] An array of hashes each representing a CDN Connect object.
    def files
      return get_result('files')
    end

    
    ##
    # Can be either a file or folder object, or the first file in the files array.
    #
    # @return [hash] A hash representing a CDN Connect object.
    def object
      obj = get_result('object')
      if obj != nil
        return obj
      end
      fs = files
      if fs != nil and fs.length > 0
        return fs[0]
      end
      nil
    end
    

    ##
    # Content body of the http response. This will be read by the
    # data method so the body can be parsed into a hash.
    #
    # @return [String]
    def body
      if @http_response != nil 
        @body = @http_response.body
      end
      @body
    end


    ## 
    # Decode the response body from JSON into a hash.
    # Store the parsed data in an instance variable (@data) so we 
    # don't keep parsing it every time we reference data.
    #
    # @return [Hash]
    def data
      if @data == nil and body != nil
        begin
          @data = JSON.parse(body)
        rescue
        end
      end
      @data
    end


    ##
    # Helper method to read the results hash.
    #
    # @return [Hash]
    def results
      d = data
      if d != nil and d.has_key?('results')
        return d['results']
      end
      nil
    end


    ##
    # A helper method to read a key within the results.
    def get_result(key)
      r = results
      if r != nil
        return r.fetch(key, nil)
      end
      nil
    end


    ##
    # An array of messages, and each message is a hash.
    # Example message within the msgs array: 
    # "text" => "info about the message", "status" => "error"
    #
    # @return [array] A response object with helper methods to read the response.
    def msgs
      d = data
      if d != nil
        return d.fetch('msgs', [])
      end
      return []
    end
    

    ##
    # @return [bool] A boolean indicating if this response has messages or not.
    def has_msgs
      msgs and msgs.length > 0
    end
    

    ##
    # @return [bool] A boolean indicating if this response has messages or not.
    def has_errors
      if msgs and msgs.length > 0
        for msg in msgs
          if msg["status"] == "error"
            return true
          end
        end
      end
      false
    end


    ##
    # Used to merge two responses into one APIResponse. This is
    # done automatically when uploading numerous files.
    #
    # @return [APIResponse]
    def merge(updating_response)
      if updating_response == nil or updating_response.data == nil
        return self
      end
      
      if updating_response.status > status
        @status = updating_response.status
      end
      
      if data == nil
        @data = updating_response.data
        return self
      end
      
      if updating_response.body != nil
        if @body == nil
          @body = updating_response.body
        else
          @body += updating_response.body
        end
      end

      @data['msgs'] += updating_response.msgs
      
      if updating_response.files != nil
          f = files
          if f != nil
            @data['results']['files'] += updating_response.files
          else
            @data['results']['files'] = updating_response.files
          end
      end
      
      return self
    end


    ##
    # @return [int] The HTTP response status.
    def status
      @status
    end

    ##
    # @return [bool]
    def is_success
      @body != nil and status >= 200 and status < 300 and not has_errors
    end

    ##
    # @return [bool]
    def is_error
      status >= 400 or has_errors
    end

    ##
    # @return [bool]
    def is_client_error
      status >= 400 and status < 500
    end

    ##
    # @return [bool]
    def is_bad_request
      status == 400
    end

    ##
    # @return [bool]
    def is_unauthorized
      status == 401
    end

    ##
    # @return [bool]
    def is_forbidden
      status == 403
    end

    ##
    # @return [bool]
    def is_not_found
      status == 404
    end

    ##
    # @return [bool]
    def is_method_not_allowed
      status == 405
    end

    ##
    # @return [bool]
    def is_server_error
      status >= 500
    end

    ##
    # @return [bool]
    def is_bad_gateway
      status == 502
    end

    ##
    # @return [bool]
    def is_service_unavailable
      status == 503
    end

    def to_s # called with print / puts
      "Status: #{status}\nBody: #{body}"
    end

  end

end